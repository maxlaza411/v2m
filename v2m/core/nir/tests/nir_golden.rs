use std::path::PathBuf;

use v2m_formats::load_nir;
use v2m_nir::normalize_nir;

fn data_dir() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../../tests/golden/nir")
}

fn load_normalized(name: &str) -> v2m_nir::NormalizedNir {
    let path = data_dir().join(format!("{name}.nir.json"));
    let nir = load_nir(&path).expect("load nir design");
    normalize_nir(&nir).expect("normalize design")
}

fn assert_snapshot(name: &str) {
    let mut settings = insta::Settings::new();
    settings.set_snapshot_path(data_dir());
    settings.set_snapshot_suffix("snap.json");
    settings.set_omit_expression(true);
    settings.set_prepend_module_to_snapshot(false);
    let _guard = settings.bind_to_scope();

    let design = load_normalized(name);
    insta::assert_json_snapshot!(name, design);
}

#[test]
fn golden_fa1() {
    assert_snapshot("fa1");
}

#[test]
fn golden_alu4() {
    assert_snapshot("alu4");
}

#[test]
fn golden_reg1x8() {
    assert_snapshot("reg1x8");
}
