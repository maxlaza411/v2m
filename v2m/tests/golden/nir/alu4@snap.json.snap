---
source: core/nir/tests/nir_golden.rs
assertion_line: 25
---
{
  "design": "alu4",
  "top": "alu4",
  "modules": {
    "alu4": {
      "inputs": {
        "a": [
          {
            "node": 0
          },
          {
            "node": 1
          },
          {
            "node": 2
          },
          {
            "node": 3
          }
        ],
        "b": [
          {
            "node": 4
          },
          {
            "node": 5
          },
          {
            "node": 6
          },
          {
            "node": 7
          }
        ],
        "op": [
          {
            "node": 8
          },
          {
            "node": 9
          }
        ]
      },
      "outputs": {
        "y": [
          {
            "node": 58
          },
          {
            "node": 61
          },
          {
            "node": 64
          },
          {
            "node": 67
          }
        ]
      },
      "states": [],
      "nodes": [
        {
          "width": 1,
          "id": 0,
          "kind": {
            "kind": "input"
          }
        },
        {
          "width": 1,
          "id": 1,
          "kind": {
            "kind": "input"
          }
        },
        {
          "width": 1,
          "id": 2,
          "kind": {
            "kind": "input"
          }
        },
        {
          "width": 1,
          "id": 3,
          "kind": {
            "kind": "input"
          }
        },
        {
          "width": 1,
          "id": 4,
          "kind": {
            "kind": "input"
          }
        },
        {
          "width": 1,
          "id": 5,
          "kind": {
            "kind": "input"
          }
        },
        {
          "width": 1,
          "id": 6,
          "kind": {
            "kind": "input"
          }
        },
        {
          "width": 1,
          "id": 7,
          "kind": {
            "kind": "input"
          }
        },
        {
          "width": 1,
          "id": 8,
          "kind": {
            "kind": "input"
          }
        },
        {
          "width": 1,
          "id": 9,
          "kind": {
            "kind": "input"
          }
        },
        {
          "width": 1,
          "id": 10,
          "kind": {
            "kind": "const",
            "bits": "0"
          }
        },
        {
          "width": 1,
          "id": 11,
          "kind": {
            "kind": "op",
            "op": "XOR",
            "inputs": [
              {
                "node": 0
              },
              {
                "node": 4
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 12,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 0
              },
              {
                "node": 4
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 13,
          "kind": {
            "kind": "op",
            "op": "XOR",
            "inputs": [
              {
                "node": 1
              },
              {
                "node": 5
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 14,
          "kind": {
            "kind": "op",
            "op": "XOR",
            "inputs": [
              {
                "node": 12
              },
              {
                "node": 13
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 15,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 1
              },
              {
                "node": 5
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 16,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 12
              },
              {
                "node": 13
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 17,
          "kind": {
            "kind": "op",
            "op": "OR",
            "inputs": [
              {
                "node": 15
              },
              {
                "node": 16
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 18,
          "kind": {
            "kind": "op",
            "op": "XOR",
            "inputs": [
              {
                "node": 2
              },
              {
                "node": 6
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 19,
          "kind": {
            "kind": "op",
            "op": "XOR",
            "inputs": [
              {
                "node": 17
              },
              {
                "node": 18
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 20,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 2
              },
              {
                "node": 6
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 21,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 17
              },
              {
                "node": 18
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 22,
          "kind": {
            "kind": "op",
            "op": "OR",
            "inputs": [
              {
                "node": 20
              },
              {
                "node": 21
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 23,
          "kind": {
            "kind": "op",
            "op": "XOR",
            "inputs": [
              {
                "node": 3
              },
              {
                "node": 7
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 24,
          "kind": {
            "kind": "op",
            "op": "XOR",
            "inputs": [
              {
                "node": 22
              },
              {
                "node": 23
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 25,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 3
              },
              {
                "node": 7
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 26,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 22
              },
              {
                "node": 23
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 27,
          "kind": {
            "kind": "op",
            "op": "OR",
            "inputs": [
              {
                "node": 25
              },
              {
                "node": 26
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 28,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 8,
                "inverted": true
              },
              {
                "node": 11
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 29,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 8
              },
              {
                "node": 12
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 30,
          "kind": {
            "kind": "op",
            "op": "OR",
            "inputs": [
              {
                "node": 28
              },
              {
                "node": 29
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 31,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 8,
                "inverted": true
              },
              {
                "node": 14
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 32,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 8
              },
              {
                "node": 15
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 33,
          "kind": {
            "kind": "op",
            "op": "OR",
            "inputs": [
              {
                "node": 31
              },
              {
                "node": 32
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 34,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 8,
                "inverted": true
              },
              {
                "node": 19
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 35,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 8
              },
              {
                "node": 20
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 36,
          "kind": {
            "kind": "op",
            "op": "OR",
            "inputs": [
              {
                "node": 34
              },
              {
                "node": 35
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 37,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 8,
                "inverted": true
              },
              {
                "node": 24
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 38,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 8
              },
              {
                "node": 25
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 39,
          "kind": {
            "kind": "op",
            "op": "OR",
            "inputs": [
              {
                "node": 37
              },
              {
                "node": 38
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 40,
          "kind": {
            "kind": "op",
            "op": "OR",
            "inputs": [
              {
                "node": 0
              },
              {
                "node": 4
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 41,
          "kind": {
            "kind": "op",
            "op": "OR",
            "inputs": [
              {
                "node": 1
              },
              {
                "node": 5
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 42,
          "kind": {
            "kind": "op",
            "op": "OR",
            "inputs": [
              {
                "node": 2
              },
              {
                "node": 6
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 43,
          "kind": {
            "kind": "op",
            "op": "OR",
            "inputs": [
              {
                "node": 3
              },
              {
                "node": 7
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 44,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 8,
                "inverted": true
              },
              {
                "node": 40
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 45,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 8
              },
              {
                "node": 11
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 46,
          "kind": {
            "kind": "op",
            "op": "OR",
            "inputs": [
              {
                "node": 44
              },
              {
                "node": 45
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 47,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 8,
                "inverted": true
              },
              {
                "node": 41
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 48,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 8
              },
              {
                "node": 13
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 49,
          "kind": {
            "kind": "op",
            "op": "OR",
            "inputs": [
              {
                "node": 47
              },
              {
                "node": 48
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 50,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 8,
                "inverted": true
              },
              {
                "node": 42
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 51,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 8
              },
              {
                "node": 18
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 52,
          "kind": {
            "kind": "op",
            "op": "OR",
            "inputs": [
              {
                "node": 50
              },
              {
                "node": 51
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 53,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 8,
                "inverted": true
              },
              {
                "node": 43
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 54,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 8
              },
              {
                "node": 23
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 55,
          "kind": {
            "kind": "op",
            "op": "OR",
            "inputs": [
              {
                "node": 53
              },
              {
                "node": 54
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 56,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 9,
                "inverted": true
              },
              {
                "node": 30
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 57,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 9
              },
              {
                "node": 46
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 58,
          "kind": {
            "kind": "op",
            "op": "OR",
            "inputs": [
              {
                "node": 56
              },
              {
                "node": 57
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 59,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 9,
                "inverted": true
              },
              {
                "node": 33
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 60,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 9
              },
              {
                "node": 49
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 61,
          "kind": {
            "kind": "op",
            "op": "OR",
            "inputs": [
              {
                "node": 59
              },
              {
                "node": 60
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 62,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 9,
                "inverted": true
              },
              {
                "node": 36
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 63,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 9
              },
              {
                "node": 52
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 64,
          "kind": {
            "kind": "op",
            "op": "OR",
            "inputs": [
              {
                "node": 62
              },
              {
                "node": 63
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 65,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 9,
                "inverted": true
              },
              {
                "node": 39
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 66,
          "kind": {
            "kind": "op",
            "op": "AND",
            "inputs": [
              {
                "node": 9
              },
              {
                "node": 55
              }
            ]
          }
        },
        {
          "width": 1,
          "id": 67,
          "kind": {
            "kind": "op",
            "op": "OR",
            "inputs": [
              {
                "node": 65
              },
              {
                "node": 66
              }
            ]
          }
        }
      ]
    }
  }
}
